src/category_theory/monad/bundled.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2020 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import category_theory.monad.basic
+import category_theory.eq_to_hom
+
+/-!
+# Bundled Monads
+
+We define bundled (co)monads as a structure consisting of a functor `func : C ⥤ C` endowed with
+a term of type `(co)monad func`. See `category_theory.monad.basic` for the definition.
+The type of bundled (co)monads on a category `C` is denoted `(Co)Monad C`.
+
+We also define morphisms of bundled (co)monads as morphisms of their underlying (co)monads
+in the sense of `category_theory.(co)monad_hom`. We construct a category instance on `(Co)Monad C`.
+-/
+
+namespace category_theory
+open category
+
+universes v u -- declare the `v`'s first; see `category_theory.category` for an explanation
+
+variables (C : Type u) [category.{v} C]
+
+/-- Bundled monads. -/
+structure Monad :=
+(func : C ⥤ C)
+(str : monad func . tactic.apply_instance)
+
+/-- Bundled comonads -/
+structure CoMonad :=
+(func : C ⥤ C)
+(str : comonad func . tactic.apply_instance)
+
+namespace Monad
+
+/-- The initial monad. TODO: Prove it's initial. -/
+def initial : Monad C := { func := 𝟭 _ }
+
+variable {C}
+
+instance : inhabited (Monad C) := ⟨initial C⟩
+
+instance {M : Monad C} : monad M.func := M.str
+
+/-- Morphisms of bundled monads. -/
+def hom (M N : Monad C) := monad_hom M.func N.func
+
+namespace hom
+instance {M : Monad C} : inhabited (hom M M) := ⟨monad_hom.ident _⟩
+end hom
+
+instance : category (Monad C) :=
+{ hom := hom,
+  id := λ _, monad_hom.ident _,
+  comp := λ _ _ _, monad_hom.comp,
+  id_comp' := λ _ _, by apply monad_hom.ident_comp,
+  comp_id' := λ _ _, by apply monad_hom.comp_ident,
+  assoc' := λ _ _ _ _, by apply monad_hom.comp_assoc }
+
+/-- The forgetful functor from `Monad C` to `C ⥤ C`. -/
+def forget : Monad C ⥤ (C ⥤ C) :=
+{ obj := func,
+  map := λ _ _ f, f.to_nat_trans }
+
+@[simp]
+lemma comp_to_nat_trans {M N L : Monad C} (f : M ⟶ N) (g : N ⟶ L) :
+  (f ≫ g).to_nat_trans = nat_trans.vcomp f.to_nat_trans g.to_nat_trans := rfl
+
+theorem hext (M N : Monad C) : M.func = N.func → (η_ M.func) == (η_ N.func) →
+  (μ_ M.func) == (μ_ N.func) → M = N :=
+begin
+  intros h1 h2 h3,
+  cases M, cases N,
+  dsimp only [] at h1,
+  subst h1,
+  congr,
+  cases M_str, cases N_str,
+  congr,
+  repeat {apply eq_of_heq, assumption}
+end
+
+@[simp] lemma assoc_func_app {M : Monad C} {X : C} :
+  M.func.map ((μ_ M.func).app X) ≫ (μ_ M.func).app X =
+  (μ_ M.func).app (M.func.obj X) ≫ (μ_ M.func).app X := by apply monad.assoc
+
+end Monad
+
+namespace CoMonad
```

Comment:
I'm not a huge fan of `CoMonad`, I'd prefer `Comonad`, but this is definitely just my personal taste

Reply:
I don't really have any strong opinions about this, although I do pronounce "comonad" as "co-monad" so "CoMonad" makes sense to me.

Reply:
I think I'm with @b-mehta here, although there's little point bikeshedding it as Lean4 wants us all to change to CamelCase, so we have some renaming in the future anyway!

Reply:
Okay, I renamed "CoMonad" to "Comonad".